openedx_learning/core/components/api.py

```diff
@@ -18,44 +18,48 @@
 from django.db.models import Q, QuerySet
 from django.db.transaction import atomic
 
-from ...lib.cache import lru_cache
 from ..publishing import api as publishing_api
-from .models import Component, ComponentType, ComponentVersion, ComponentVersionRawContent
+from .models import Component, ComponentType, ComponentVersion, ComponentVersionContent
 
 
-@lru_cache(maxsize=128)
-def get_or_create_component_type_id(namespace: str, name: str) -> int:
+def get_or_create_component_type(namespace: str, name: str) -> ComponentType:
     """
     Get the ID of a ComponentType, and create if missing.
+
+    Caching Warning: Be careful about putting any caching decorator around this
+    function (e.g. ``lru_cache``). It's possible that incorrect cache values
+    could leak out in the event of a rollback–e.g. new types are introduced in
+    a large import transaction which later fails. You can safely cache the
+    results that come back from this function with a local dict in your import
+    process instead.#
     """
     component_type, _created = ComponentType.objects.get_or_create(
         namespace=namespace,
         name=name,
     )
-    return component_type.id
+    return component_type
 
 
 def create_component(
     learning_package_id: int,
     /,
-    namespace: str,
-    type_name: str,
+    component_type: ComponentType,
```

Comment:
I'd rather keep this as is for now. The main reason I made `create_component` take a `ComponentType` instead of an ID or the namespace/name combo was so that `create_component` wouldn't have to query for `ComponentType` information every time. I figured that we often create many Components at once, like we would for an import, and that the caller can take the hit for pulling the ComponentType once per type.

But the `get_component_by_key` and `component_exists_by_key` is fetching one component, usually as a lookup for rendering a request for a specific block (since we have much more efficient ways of iterating through lists of them). So I thought it'd be easier for them to just pass through the arguments and let the API worry about fetching them in one query, and not have the caller think about `ComponentTypes`.

I realize that there are a few different ways we could add caching to the mix to make this better, but they all carry complications that I didn't think were worth it just yet (e.g. request caching adds a major dependency and can have edge cases around celery tasks, `lru` has the rollback invalidation edge case, etc.).

I'd probably just make a function for both if Python had better support for it. (I mean, I know `singledispatch` exists, but I've never seen anyone actually use it.)

Reply:
Makes sense to me 👍🏻